cty/function/stdlib/format_test.go

```diff
@@ -836,6 +836,15 @@ func TestFormatList(t *testing.T) {
 			cty.UnknownVal(cty.List(cty.String)),
 			``,
 		},
+		23: {
+			cty.StringVal("%v"),
+			[]cty.Value{cty.DynamicVal},
+			// The current Function implementation will default to DynamicVal
+			// if AllowUnknown is true, even though this function has a static
+			// return type
+			cty.DynamicVal,
```

Comment:
I'm kinda surprised to see this work at all because (as I think your comment here is noting) `Function.Call` is supposed to guarantee to return a value that conforms to whatever `Function.ReturnTypeForValues` would've returned given the same arguments, and `cty.DynamicVal` doesn't conform to `cty.List(cty.String)`. :thinking: 

https://github.com/zclconf/go-cty/blob/aba65fbeb4f9726cbee4c59754cf3d2b95629737/cty/function/function.go#L304-L311

I'm fine with leaving that bug to be solved in a separate PR, but was wondering if you did any investigation here that revealed why exactly this case returns the wrong type. If you don't know then I'm happy to go digging for that myself, but asking just in case you already did the research here. :grinning:

Reply:
Nope, I didn't dig into the `Function.Call` implementation yet. If I recall when I read through it for this bug, it looked like returning `cty.DynamicVal` was simply the easy path, but there was room for more accurate type checking which we could evaluated separately.